### ISSUE TITLE
PDF document stays in use after closing

### ISSUE THREAD

**Issue by shashikulk** [pymupdf/PyMuPDF#3140]
I want to open an existing PDF file, update it and save it with a new name. Once the updated PDF is saved with a new name, I want to delete the existing / old PDF file. I am using ```insert_htmlbox()``` as one of the functions to update the PDF. My observation is that due to ```insert_htmlbox()``` the old file is remaining open and hence the ```os.remove()``` on old file is giving error. 

Below is the code snippet I am using. PyMuPDF version = 1.23.21

```
import fitz
import os

oldfile = 'your-existing-or-old-filename.pdf'
newfile = 'your-new-filename.pdf'
css2 = """
@font-face {font-family: QuicksandFM; src: url(Quicksand-Regular.ttf);}
@font-face {font-family: QuicksandFM; src: url(Quicksand-Bold.ttf); font-weight: bold;} 
body {font-family:QuicksandFM; font-size:9;}
"""

with fitz.open(oldfile) as doc:  # open document
    page = doc[0]
    rect = fitz.Rect(130, 400, 430, 600)
    CELLS = fitz.make_table(rect, cols=3, rows=5)
    shape = page.new_shape()  # create Shape
    for i in range(5):
        for j in range(3):
            qtext = "<b>" + "Ques #" + str(i*3+j+1) + ": " + "</b>"
            atext = "<b>" + "Ans:" + "</b>"
            qtext = qtext + '<br>' + atext
            shape.draw_rect(CELLS[i][j])  # draw rectangle
            page.insert_htmlbox(CELLS[i][j], qtext, css=css2, scale_low=0)
    shape.finish(width=2.5, color=fitz.pdfcolor["blue"], )
    shape.commit()  # write all stuff to the page
    doc.subset_fonts()
    doc.ez_save(newfile)
os.remove(oldfile)

```
Can this be fixed?

**Comment by JorjMcKie** [comment]
You should simply add `page = None` immediately after `shape.commit()`.
This should solve your problem.

**Comment by JorjMcKie** [comment]
> Adding `page = None` after `shape.commit()` didn't help. Still getting same error.

You are right, just checked it.
Please also set `doc=None` before `os.remove`. We will investigate this again - there was a related fix recently.

**Comment by JorjMcKie** [comment]
Took the liberty to change the issue title, because this presumably is not in any way connected to `insert_htmlbox`.

**Comment by shashikulk** [comment]
Tried ```doc=None``` as well as you suggested. That too didn't work. 

> Took the liberty to change the issue title, because this presumably is not in any way connected to insert_htmlbox.

ok, np. Will wait for an update from your end.

**Comment by cm2d** [comment]
Hello,

Having observed the same phenomenon in PyMuPDF version = 1.23.21, I take this opportunity to ask a related question:

Is it possible to unload memory as pages are read? This would facilitate the processing of large documents, for which once the elements of the page are retrieved, there is no need to keep it in memory.

However, unless I'm mistaken, there is no page.close() function like there is for doc.close().

Thank you in advance for your response.

And congratulations on the work done!

**Comment by JorjMcKie** [comment]
@cm2d if you simply do `page = None` or `del page`, the `__del__` method of the object is invoked.
This also frees resources that the Python `Page` object holds on the C-level (inside MuPDF).

If you want to free more memory, you can always invoke [store_shrink](https://pymupdf.readthedocs.io/en/latest/tools.html#Tools.store_shrink), which empties MuPDF's cache. This area is used to improve performance obviously and contains large objects primarily, like images. Its contents are not page- or document-related.

**Comment by julian-smith-artifex-com** [comment]
Fixed in 1.23.23.
